## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [32]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [33]:
# Write your code for importing the csv file here
movies = pd.read_csv("Movies.csv")
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [5]:
# Write your code for inspection here
movies.shape


(3853, 28)

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [14]:
movies.isnull().sum()

color                         2
director_name                 0
num_critic_for_reviews        1
duration                      1
director_facebook_likes       0
actor_3_facebook_likes        6
actor_2_name                  1
actor_1_facebook_likes        0
gross                         0
genres                        0
actor_1_name                  0
movie_title                   0
num_voted_users               0
cast_total_facebook_likes     0
actor_3_name                  6
facenumber_in_poster          6
plot_keywords                30
movie_imdb_link               0
num_user_for_reviews          0
language                      3
country                       0
content_rating               48
budget                        0
title_year                    0
actor_2_facebook_likes        1
imdb_score                    0
aspect_ratio                 72
movie_facebook_likes          0
dtype: int64

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [34]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(['color',
'director_facebook_likes',
'actor_1_facebook_likes',
'actor_2_facebook_likes',
'actor_3_facebook_likes',
'actor_2_name',
'cast_total_facebook_likes',
'actor_3_name',
'duration',
'facenumber_in_poster',
'content_rating',
'country',
'movie_imdb_link',
'aspect_ratio',
'plot_keywords'], inplace=True, axis=1)


In [21]:
len(movies.columns)

13

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [ ]:
# Write you code here


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [27]:
# Write your code for filling the NaN values in the 'language' column here
movies[movies['language'].isnull()]
movies[movies['language'] == 'English']

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3844        John Waters                    73.0     180483.0   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                               genres     actor_1_name  \
0     Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1            Action|Adventure|Fantasy      Johnny Depp   
2           Action|Adventure|Thriller  Christoph Waltz   
3                     Action|Thriller        Tom Hardy   
4             Action|Adventure|Sci-Fi     Daryl Sabara   
...                               ...              ...   
3844              Comedy|Crime|Horror           Divine   
3848            Drama|Sci-Fi|Thriller    Shane Carruth   
3849                         Thriller      Ian Gamazon   
3851                     Comedy|Drama      Kerry Bishé   
3852                      Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3844                            Pink Flamingos             16792   
3848                                    Primer             72639   
3849                                    Cavite               589   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3844                 183.0  English      10000.0      1972.0         6.1   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3844                     0  
3848                 19000  
3849                    74  
3851                   413  
3852                   456  

[3671 rows x 13 columns]

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [ ]:
# Write your code for unit conversion here


**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [ ]:
# Write your code for creating the profit column here


In [ ]:
# Write your code for sorting the dataframe here


In [ ]:
top10 = # Write your code to get the top 10 profiting movies here

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [47]:
movies.head()
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] == 'Meryl Streep']

In [48]:
Leo_Caprio = movies[movies['actor_1_name'] == 'Leonardo DiCaprio']

In [49]:
Brad_Pitt = movies[movies['actor_1_name'] == 'Brad Pitt']
Brad_Pitt.columns.str.capitali

In [50]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])
Combined.head()

director_name  num_critic_for_reviews        gross  \
392     Nancy Meyers                   187.0  112703470.0   
1038   Curtis Hanson                    42.0   46815748.0   
1132     Nora Ephron                   252.0   94125426.0   
1322   David Frankel                   208.0  124732962.0   
1390  Robert Redford                   227.0   14998070.0   

                               genres  actor_1_name             movie_title  \
392              Comedy|Drama|Romance  Meryl Streep       It's Complicated    
1038  Action|Adventure|Crime|Thriller  Meryl Streep         The River Wild    
1132          Biography|Drama|Romance  Meryl Streep          Julie & Julia    
1322             Comedy|Drama|Romance  Meryl Streep  The Devil Wears Prada    
1390               Drama|Thriller|War  Meryl Streep        Lions for Lambs    

      num_voted_users  num_user_for_reviews language      budget  title_year  \
392             69860                 214.0  English  85000000.0      2009.0   
1038            32544                  69.0  English  45000000.0      1994.0   
1132            79264                 277.0  English  40000000.0      2009.0   
1322           286178                 631.0  English  35000000.0      2006.0   
1390            41170                 298.0  English  35000000.0      2007.0   

      imdb_score  movie_facebook_likes  
392          6.6                     0  
1038         6.3                     0  
1132         7.0                 13000  
1322         6.8                     0  
1390         6.2                     0

In [51]:
# Write your code for grouping the combined dataframe here
actor_groupby = Combined.groupby(by='actor_1_name').mean()

In [52]:
# Write the code for finding the mean of critic reviews and audience reviews here
actor_groupby

num_critic_for_reviews         gross  num_voted_users  \
actor_1_name                                                               
Brad Pitt                      245.000000  6.632145e+07    283583.823529   
Leonardo DiCaprio              330.190476  1.257420e+08    441958.857143   
Meryl Streep                   181.454545  5.991973e+07     73545.545455   

                   num_user_for_reviews        budget   title_year  \
actor_1_name                                                         
Brad Pitt                    742.352941  6.929412e+07  2004.352941   
Leonardo DiCaprio            914.476190  8.245238e+07  2005.476190   
Meryl Streep                 297.181818  3.445455e+07  2003.545455   

                   imdb_score  movie_facebook_likes  
actor_1_name                                         
Brad Pitt            7.152941          14051.764706  
Leonardo DiCaprio    7.495238          52190.476190  
Meryl Streep         6.745455           2934.090909

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [53]:
import pandas as pd

In [56]:
s1 = pd.Series([8, 1,14,0,7,5,6,7,13,9,10])

In [57]:
s2 = pd.Series([0, 1,2,3,4,5,6,7,8,9,10])

In [63]:
s1 not in s2

TypeError: unhashable type: 'Series'

In [62]:
import pandas as pd 

# Reading the input dataframe
pima = pd.read_csv('https://media-doselect.s3.amazonaws.com/generic/pLZK3n22ezVwAG2XOYW5qEx7V/pima_indian_diabetes.csv')
pima['BMI'] = round(pima['BMI'])
# Write your code here
diabatic = pima.pivot_table(values=['Diabetes'], index='BMI', aggfunc='sum')
diabatic.sort_values(by='Diabetes', inplace=True, ascending=False)
diabatic

Diabetes
BMI           
34.0        15
38.0        10
36.0        10
30.0        10
35.0         9
33.0         9
32.0         8
37.0         7
28.0         5
43.0         5
46.0         5
39.0         4
42.0         4
31.0         4
27.0         3
29.0         3
26.0         3
41.0         3
44.0         2
23.0         2
40.0         2
50.0         1
53.0         1
55.0         1
45.0         1
59.0         1
67.0         1
25.0         1
19.0         0
47.0         0
48.0         0
24.0         0
52.0         0
22.0         0
21.0         0
57.0         0
20.0         0
18.0         0

In [73]:
# Write your code for importing the csv file here
movies = pd.read_csv("Movies.csv")
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

In [65]:
movies['aspect_ratio'].isnull().sum()

72

In [66]:
movies['facenumber_in_poster'].isnull().sum()

6

In [74]:
# Importing the pandas package
import pandas as pd 

# Reading the movies dataframe
movies = pd.read_csv('https://media-doselect.s3.amazonaws.com/generic/1M2ZzY2M9PEBPJovgoaBgZdbM/movie_data.csv')

# Your aim is to complete the following three print statements after all the colons

mv = {'aspect_ratio_mv': movies['aspect_ratio'].isnull().sum(), 'facenumber_in_poster_mv': movies['facenumber_in_poster'].isnull().sum()}
median = {'aspect_ratio_median': movies['aspect_ratio'].median(), 'facenumber_in_poster_median': movies['facenumber_in_poster'].median()}
movies['aspect_ratio'].fillna(movies['aspect_ratio'].median())
movies['facenumber_in_poster'].fillna(movies['facenumber_in_poster'].median())
final = {'aspect_ratio_final': movies['aspect_ratio'].median(), 'facenumber_in_poster_final': movies['facenumber_in_poster'].median()}

# Printing the values in the three dictionaries. Please do not edit this part
print(sorted(mv.values()))
print(sorted(median.values()))
print(sorted(final.values()))

[13, 329]
[1.0, 2.35]
[1.0, 2.35]


In [75]:

movies.loc([(movies['aspect_ratio'].isnull())], 'aspect_ratio') =  movies['aspect_ratio'].median()
movies.loc([(movies['aspect_ratio'].isnull())], 'aspect_ratio')

SyntaxError: cannot assign to function call (Temp/ipykernel_1208/2563639392.py, line 1)

There are two columns 'actor_1_facebook_likes' and 'actor_2_facebook_likes' which have quite a few missing values but since you have less data points, you don't want to drop many of them. For a hypothetical analysis that you're conducting, it will be okay if one of them has a missing value but you can't afford to have both missing values. So your aim here is to find the indices of the rows in which both of these columns have missing values simultaneously.

Expected Output:
- First print the indices of the rows where both these columns have missing values. The print statement has been provided in the stub. You just need to fill it. 
- After you have printed the above indices, drop these particular rows and print the number of retained rows in the dataframe.

A sample output would look like the following:
[389, 1019, 1178, 3400, 4012]
4847

Here, the list in the first line indicates a sample list which indicates the indices of the rows where both of the columns have missing values. And the second line represents the number of rows remaining in the dataframe after you have dropped the above rows.

In [95]:
# Importing the pandas package
import pandas as pd 

# Reading the dataframe
movies = pd.read_csv('movie_data (1).csv')
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [100]:
movies[movies[ 'actor_1_facebook_likes'].isnull() & movies[ 'actor_2_facebook_likes'].isnull()].index

Int64Index([4502, 4519, 4720, 4837, 4945, 4946, 4990], dtype='int64')

In [102]:
movies_final = pd.read_csv('movies_final (2).csv', sep='\t')

In [104]:
movies_final.head()

Unnamed: 0  color      director_name  num_critic_for_reviews  duration  \
0           0  Color      James Cameron                   723.0     178.0   
1           1  Color     Gore Verbinski                   302.0     169.0   
2           2  Color         Sam Mendes                   602.0     148.0   
3           3  Color  Christopher Nolan                   813.0     164.0   
4           5  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross  ... num_user_for_reviews language  \
0                  1000.0  760505847.0  ...               3054.0  English   
1                 40000.0  309404152.0  ...               1238.0  English   
2                 11000.0  200074175.0  ...                994.0  English   
3                 27000.0  448130642.0  ...               2701.0  English   
4                   640.0   73058679.0  ...                738.0  English   

  country  content_rating       budget title_year  actor_2_facebook_likes  \
0     USA           PG-13  237000000.0     2009.0                   936.0   
1     USA           PG-13  300000000.0     2007.0                  5000.0   
2      UK           PG-13  245000000.0     2015.0                   393.0   
3     USA           PG-13  250000000.0     2012.0                 23000.0   
4     USA           PG-13  263700000.0     2012.0                   632.0   

  imdb_score aspect_ratio  movie_facebook_likes  
0        7.9         1.78                 33000  
1        7.1         2.35                     0  
2        6.8         2.35                 85000  
3        8.5         2.35                164000  
4        6.6         2.35                 24000  

[5 rows x 29 columns]

In [108]:
movies_final['total_facebook_likes'] = movies_final['actor_1_facebook_likes'] + movies_final['actor_3_facebook_likes'] + movies_final['actor_2_facebook_likes']

In [109]:
movies_final['total_facebook_likes'] 

0        2791.0
1       46000.0
2       11554.0
3       73000.0
4        1802.0
         ...   
3751      754.0
3752        5.0
3753      344.0
3754      147.0
3755      125.0
Name: total_facebook_likes, Length: 3756, dtype: float64

In [114]:
movies_final.sort_values(by=['total_facebook_likes'])

Unnamed: 0  color      director_name  num_critic_for_reviews  duration  \
3739        4977  Color    Morgan Spurlock                   193.0     100.0   
3714        4890  Color         Tom Putnam                    22.0      86.0   
3720        4915  Color        Ricki Stern                    11.0     106.0   
2174        2382  Color     Emile Ardolino                    10.0      92.0   
479          502  Color     Antony Hoffman                   145.0     106.0   
...          ...    ...                ...                     ...       ...   
2086        2269  Color       Nimród Antal                   107.0      88.0   
2641        2939  Color  John Erick Dowdle                   198.0      89.0   
3601        4592  Color        Johnny Remo                     2.0     112.0   
1155        1223  Color     David R. Ellis                   221.0      82.0   
1766        1902  Color         Adam McKay                   181.0      98.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
3739                    293.0                     0.0    Amanda Kearsan   
3714                     15.0                     0.0     Donald Austin   
3720                     15.0                     0.0  Evelyn Jefferson   
2174                     22.0                     0.0      Kyra Nichols   
479                      14.0                     0.0        Val Kilmer   
...                       ...                     ...               ...   
2086                    190.0                   163.0         Fred Ward   
2641                     66.0                  1000.0     Jay Hernandez   
3601                     74.0                   891.0       Randy Wayne   
1155                    160.0                   748.0   Andrew Fiscella   
1766                    285.0                  7000.0      Will Ferrell   

      actor_1_facebook_likes       gross  ... language country content_rating  \
3739                     0.0  11529368.0  ...  English     USA             PG   
3714                     2.0    111300.0  ...  English     USA      Not Rated   
3720                     2.0      1111.0  ...  English     USA          PG-13   
2174                     2.0   2119994.0  ...  English     USA              G   
479                      2.0  17473245.0  ...  English     USA          PG-13   
...                      ...         ...  ...      ...     ...            ...   
2086                137000.0  15988876.0  ...  English     USA          PG-13   
2641                137000.0  31691811.0  ...  English     USA              R   
3601                260000.0     96734.0  ...  English     USA          PG-13   
1155                164000.0  66466372.0  ...  English     USA              R   
1766                640000.0  84136909.0  ...  English     USA          PG-13   

          budget  title_year actor_2_facebook_likes  imdb_score aspect_ratio  \
3739     65000.0      2004.0                    0.0         7.3         1.78   
3714    225000.0      2012.0                    0.0         7.5         2.35   
3720    200000.0      2006.0                    0.0         7.7         1.66   
2174  19000000.0      1993.0                    0.0         5.7         1.85   
479   70000000.0      2000.0                    0.0         5.7         2.35   
...          ...         ...                    ...         ...          ...   
2086  27000000.0      2009.0                  459.0         5.7         2.35   
2641  12000000.0      2008.0                 1000.0         6.0         1.85   
3601   1000000.0      2012.0                  984.0         5.6         1.85   
1155  40000000.0      2009.0               137000.0         5.2         2.35   
1766  26000000.0      2004.0                 8000.0         7.2         1.85   

     movie_facebook_likes  total_facebook_likes  
3739                    0                   0.0  
3714                  801                   2.0  
3720                  246                   2.0  
2174      

In [136]:
# Importing the pandas package
import pandas as pd 

# Reading the dataframe
movies = pd.read_csv('movie_data (1).csv')
movies.head()
# Write your code here
movies['total_facebook_likes'] = movies['actor_1_facebook_likes'] + movies['actor_3_facebook_likes'] + movies['actor_2_facebook_likes']
movies.sort_values(by=['total_facebook_likes'], inplace=True, ascending=False)
movies.head(50).loc[:,['actor_1_facebook_likes', 'actor_2_facebook_likes',  'actor_3_facebook_likes', 'total_facebook_likes', 'actor_1_name','actor_2_name', 'actor_3_name']]


actor_1_facebook_likes  actor_2_facebook_likes  actor_3_facebook_likes  \
1902                640000.0                  8000.0                  7000.0   
1223                164000.0                137000.0                   748.0   
4704                260000.0                 21000.0                  2000.0   
4592                260000.0                   984.0                   891.0   
4409                260000.0                   454.0                   354.0   
4045                164000.0                  2000.0                   898.0   
2939                137000.0                  1000.0                  1000.0   
2269                137000.0                   459.0                   163.0   
2385                 87000.0                 17000.0                 16000.0   
74                   87000.0                 11000.0                  7000.0   
555                  87000.0                 13000.0                   759.0   
972                  87000.0                 11000.0                   935.0   
105                  87000.0                  2000.0                   702.0   
2189                 87000.0                  1000.0                   602.0   
4435                 87000.0                   781.0                   711.0   
2236                 87000.0                   570.0                   470.0   
97                   29000.0                 27000.0                 23000.0   
4900                 77000.0                    21.0                    16.0   
33                   40000.0                 25000.0                 11000.0   
47                   34000.0                 22000.0                 20000.0   
106                  40000.0                 25000.0                 11000.0   
4128                 40000.0                 25000.0                 11000.0   
1550                 49000.0                 22000.0                  2000.0   
134                  40000.0                 17000.0                 16000.0   
3                    27000.0                 23000.0                 23000.0   
933                  49000.0                 22000.0                   878.0   
1185                 34000.0                 23000.0                 14000.0   
761                  34000.0                 22000.0                 14000.0   
103                  34000.0                 22000.0                 14000.0   
207                  34000.0                 22000.0                 14000.0   
38                   44000.0                 15000.0                 11000.0   
2137                 34000.0                 22000.0                 14000.0   
4778                 44000.0                 15000.0                 11000.0   
397                  49000.0                 20000.0                   971.0   
3058                 40000.0                 29000.0                   472.0   
8                    26000.0                 21000.0                 19000.0   
794                  26000.0                 21000.0                 19000.0   
17                   26000.0                 21000.0                 19000.0   
289                  40000.0                 23000.0                  1000.0   
45                   26000.0                 23000.0                 14000.0   
3089                 49000.0                 13000.0                   893.0   
1233                 23000.0                 20000.0                 19000.0   
2067                 29000.0                 22000.0                 11000.0   
905                  40000.0                 19000.0                  3000.0   
589                  49000.0                  8000.0                  5000.0   
945                  40000.0                 11000.0                 10000.0   
175                  49000.0                 11000.0                   988.0   
999                  45000.0                 13000.0                   282.0   
462                  22000.0                 20000.0                 16000.0   
1278  

In [134]:
movies.head(50).loc[:,['actor_1_name','actor_2_name', 'actor_3_name']]

actor_1_name            actor_2_name           actor_3_name
1902           Darcy Donavan            Will Ferrell           Steve Carell
1223            Krista Allen         Andrew Fiscella      Shantel VanSanten
4704            Matthew Ziff           Lorraine Ziff          Michael Biehn
4592            Matthew Ziff             Randy Wayne  Raquel Elizabeth Ames
4409            Matthew Ziff              T.J. Storm             Sam Medina
4045            Krista Allen               Eric Dane          Henry Rollins
2939         Andrew Fiscella           Jay Hernandez          Dania Ramirez
2269         Andrew Fiscella               Fred Ward            Lorna Raver
2385           Jimmy Bennett      Chloë Grace Moretz          Ryan Reynolds
74             Jimmy Bennett          Morgan Freeman           Steve Carell
555            Jimmy Bennett            Bruce Willis       Ransford Doherty
972            Jimmy Bennett           Harrison Ford      Mary Lynn Rajskub
105            Jimmy Bennett              Mike Vogel         Andre Braugher
2189           Jimmy Bennett             CCH Pounder        Aryana Engineer
4435           Jimmy Bennett       Joey Lauren Adams       Matthew Lawrence
2236           Jimmy Bennett              Leo Howard             Jake Short
97         Leonardo DiCaprio               Tom Hardy   Joseph Gordon-Levitt
4900          Michael Joiner            Chris Thomas   Michael Higgenbottom
33               Johnny Depp            Alan Rickman          Anne Hathaway
47         Jennifer Lawrence          Peter Dinklage           Hugh Jackman
106              Johnny Depp            Alan Rickman          Anne Hathaway
4128             Johnny Depp            Alan Rickman          Anne Hathaway
1550          Robin Williams          Robert De Niro           Topher Grace
134              Johnny Depp      Chloë Grace Moretz        Christopher Lee
3                  Tom Hardy          Christian Bale   Joseph Gordon-Levitt
933           Robin Williams  Philip Seymour Hoffman          Monica Potter
1185       Jennifer Lawrence          Christian Bale         Bradley Cooper
761        Jennifer Lawrence          Robert De Niro         Bradley Cooper
103        Jennifer Lawrence  Philip Seymour Hoffman        Josh Hutcherson
207        Jennifer Lawrence  Philip Seymour Hoffman        Josh Hutcherson
38                Tim Holmes              Mila Kunis           James Franco
2137       Jennifer Lawrence          Robert De Niro         Bradley Cooper
4778              Tim Holmes              Mila Kunis           James Franco
397           Robin Williams            Hugh Jackman        Elizabeth Daily
3058             Johnny Depp       Leonardo DiCaprio            Kevin Tighe
8            Chris Hemsworth       Robert Downey Jr.     Scarlett Johansson
794          Chris Hemsworth       Robert Downey Jr.     Scarlett Johansson
17           Chris Hemsworth       Robert Downey Jr.     Scarlett Johansson
289              Johnny Depp          Christian Bale         Stephen Graham
45             Jason Statham             Paul Walker             Vin Diesel
3089          Robin Williams              Matt Damon          Minnie Driver
1233          Christian Bale            Hugh Jackman     Scarlett Johansson
2067       Leonardo DiCaprio          Robert De Niro           Meryl Streep
905              Johnny Depp    Benedict Cumberbatch             Adam Scott
589           Robin Williams           Julia Roberts            Bob Hoskins
945              Johnny Depp            Meryl Streep          Anna Kendrick
175           Robin Williams               Brad Pitt                 Common
999              Muse Watson      Sylvester Stallone            Kelly Rowan
462   Philip Seymour Hoffman         Natalie Portman         Charlie Hunnam
1278         Natalie Portman      Scarlett Johansson   Benedict Cumberbatch

In [127]:
# Importing the pandas package
import pandas as pd 

# Reading the dataframe
movies = pd.read_csv('movies_final (2).csv', sep='\t')


# Write your code here
group=movies.pivot_table(values=['actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes'], index=['actor_1_name','actor_2_name','actor_3_name'],aggfunc='sum')

group['Total likes']=group['actor_1_facebook_likes']+group['actor_2_facebook_likes']+group['actor_3_facebook_likes']

group.sort_values(by=['Total likes'],inplace=True,ascending=False)

group.reset_index(inplace=True)

j=0 

for i in group['Total likes']:
    temp=sorted([group.loc[j,'actor_1_facebook_likes'],group.loc[j,'actor_2_facebook_likes'],group.loc[j,'actor_3_facebook_likes']])
    if temp[0]>=temp[1]/2 and temp[0]>=temp[2]/2 and temp[1]>=temp[2]/2:
        print(sorted([group.loc[j,'actor_1_name'],group.loc[j,'actor_2_name'],group.loc[j,'actor_3_name']]))
        print(temp)
        break
    j+=1

['Chris Hemsworth', 'Robert Downey Jr.', 'Scarlett Johansson']
[57000.0, 63000.0, 78000.0]
